palette/build/lut.rs

```diff
@@ -0,0 +1,592 @@
+    }
+    #[cfg(feature = "prophoto_lut")]
+    for entry in &entries_u16 {
+        write!(writer, ", {}", entry.fn_type).unwrap();
+    }
+    writeln!(writer, "}};").unwrap();
+
+    gen_into_linear_lut_u8(writer, &entries_u8);
+    #[cfg(feature = "prophoto_lut")]
+    gen_into_linear_lut_u16(writer, &entries_u16);
+    gen_from_linear_lut_u8(writer, &entries_u8);
+    #[cfg(feature = "prophoto_lut")]
+    gen_from_linear_lut_u16(writer, &entries_u16)
+}
+
+fn gen_into_linear_lut_u8(writer: &mut File, entries: &[LutEntryU8]) {
+    use std::io::Write;
+
+    for LutEntryU8 {
+        fn_type,
+        fn_type_uppercase,
+        into_linear,
+        ..
+    } in entries
+    {
+        let table_size = 1 << 8;
+        let mut table = Vec::new();
+        for i in 0..table_size {
+            let encoded = (i as f64) / ((table_size - 1) as f64);
+            let linear = into_linear(encoded);
+            // Handle integer floats printing without decimal
+            let float_string = if linear <= 0.0 {
+                "\t0.0,\n".to_owned()
+            } else if linear >= 1.0 {
+                "\t1.0,\n".to_owned()
+            } else {
+                format!("\t{linear},\n")
+            };
+            table.extend_from_slice(float_string.as_bytes());
+        }
+
+        let table_name = format!("{fn_type_uppercase}_U8_TO_F64");
+        writeln!(writer, "const {table_name}: [f64; {table_size}] = [").unwrap();
+        writer.write_all(&table).unwrap();
+        writeln!(writer, "];").unwrap();
+
+        writeln!(
+            writer,
+            "impl IntoLinear<f64, u8> for {fn_type} {{\
+        \n\t#[inline]\
+        \n\tfn into_linear(encoded: u8) -> f64 {{\
+        \n\t\t{table_name}[encoded as usize]\
+        \n\t}}\
+        \n}}"
+        )
+        .unwrap();
+
+        writeln!(
+            writer,
+            "impl IntoLinear<f32, u8> for {fn_type} {{\
+        \n\t#[inline]\
+        \n\tfn into_linear(encoded: u8) -> f32 {{\
+        \n\t\t{table_name}[encoded as usize] as f32\
+        \n\t}}\
+        \n}}"
+        )
+        .unwrap();
+    }
+}
+
+#[cfg(feature = "prophoto_lut")]
+fn gen_into_linear_lut_u16(writer: &mut File, entries: &[LutEntryU16]) {
+    use std::io::Write;
+
+    for LutEntryU16 {
+        fn_type,
+        fn_type_uppercase,
+        into_linear,
+        ..
+    } in entries
+    {
+        let table_size = 1 << 16;
+        let mut table = Vec::new();
+        for i in 0..table_size {
+            let encoded = (i as f64) / ((table_size - 1) as f64);
+            let linear = into_linear(encoded);
+            // Handle integer floats printing without decimal
+            let float_string = if linear <= 0.0 {
+                "\t0.0,\n".to_owned()
+            } else if linear >= 1.0 {
+                "\t1.0,\n".to_owned()
+            } else {
+                format!("\t{linear},\n")
+            };
+            table.extend_from_slice(float_string.as_bytes());
+        }
```

Comment:
I'm not sure of the overall impact, but it seems like a single string could be reused here and other places instead of some of the `format!` calls inside loops which return a new `String` every time. You can use `std::fmt::Write` to write formatted data to a buffer, but it has to be imported in a slightly funny way to avoid messing with the `std::io::Write` `write!`/`writeln!` macros. Also have to remember to clear it before reusing it.

https://doc.rust-lang.org/stable/std/macro.write.html#examples

---

Looking at the removed tables, it seems only the first and last values are `0.0` and`1.0`, respectively. If so, you could rewrite this loop to make fewer allocations and avoid all the clamping checks. Untested

```rust
use std::fmt::Write as _;
use std::io::Write as _;

let mut float_string = String::new();

// Handle integer floats printing without decimal
table.extend_from_slice(b"\t0.0,\n");

for i in 1..table size - 1 {
    let encoded = (i as f64) / ((table_size - 1) as f64);
    let linear = into_linear(encoded);
    
    write!(&mut float_string, "\t{linear}\n").unwrap();
    table.extend_from_slice(float_string.as_bytes());
    
    float_string.clear();
}

// Handle integer floats printing without decimal
table.extend_from_slice(b"\t1.0,\n");
```



Some possible